### PR TITLE
[docs] 4D attention masks

### LINKED ISSUE #46458: Suggestion: Give a warning for 4D attention mask which use float dtype but in 0.0 and 1.0 term (instead of -inf and zero).
### Feature request

The title said it all.

### Motivation

Let's say I have a code like this: (Early warning: If you run this code, it will give an error.)

```py
import torch
from transformers import AutoModel

model = AutoModel.from_pretrained("google-bert/bert-base-uncased", device_map="auto")
attention_mask = torch.tensor([
    [1, 0],
    [0, 0],
], device=model.device).unsqueeze(0).unsqueeze(1)

outputs = model(
    input_ids=torch.tensor([[200, 201]], device=model.device),
    attention_mask=attention_mask,
)
print("Output 1:")
print(outputs.last_hidden_state[0, 0, :10])

outputs = model(
    input_ids=torch.tensor([[200, 202]], device=model.device),
    attention_mask=attention_mask,
)
print("Output 2:")
print(outputs.last_hidden_state[0, 0, :10])
```

Notice that the last input ID for output 1 and output 2 are different: 201 and 202. However, since it's designed to be masked, then the output should not be impacted.

But if you run that code, it will return an error because the attention mask are expected to be in float or boolean type.

---

My previous attempt was converting it to float, so the line 5-8 should be changed like this:
```py
...
attention_mask = torch.tensor([
    [1, 0],
    [0, 0],
], device=model.device).unsqueeze(0).unsqueeze(1).float()
...
```

But here's the funky part: if you use that change, both outputs give different result! I was surprising when I found it, so I try to debug it.

After a debug session that leads me to a [Pytorch page](https://docs.pytorch.org/docs/2.12/generated/torch.nn.functional.scaled_dot_product_attention.html), my mistake is that it should be changed to boolean, so the change should be like this:
```py
...
attention_mask = torch.tensor([
    [1, 0],
    [0, 0],
], device=model.device).unsqueeze(0).unsqueeze(1).bool()
...
```

If I want to use in a float term, the 0 terms should be changed to negative infinity, and the 1 terms should be changed to zero. This is because the attention mask is treated as bias with addition operator.

There is no warning for this, so careless developer like me can just use the mask without knowing the side effect.

### Your contribution

I'm not sure if I can make a PR for this case because I haven't explored how to apply the warning in general. What I found right now is just a specific case, and I hope it can help.

### PR DESCRIPTION
**Pull request by stevhliu**
fixes #46458 by adding docs for how to create and pass a custom 4D attention mask

### CODE REVIEW

File: docs/source/en/attention_interface.md [hunk 1]
```diff
@@ -307,6 +307,57 @@ attention_mask = create_causal_mask(**mask_kwargs)
 
 During generation, [`~GenerationMixin.generate`] builds masks through [`create_masks_for_generate`], which dispatches to the right `create_*_mask` based on the model config. Override it on a model class to plug in a custom masking strategy for generation.
 
+## Pass a custom 4D attention mask
+
+Pass your own 4D mask when you need an attention pattern the `create_*_mask` functions can't express. A 4D mask has the shape `(batch_size, 1, query_length, kv_length)`, which Transformers detects and uses the mask as-is, skipping `create_*_mask`.
```
Comment (vasqu) [inline docs/source/en/attention_interface.md]: I would note the 1 stands for broadcasting across the attention heads (they can theoretically have different masks for each q head)

File: docs/source/en/attention_interface.md [hunk 2]
```diff
@@ -307,6 +307,57 @@ attention_mask = create_causal_mask(**mask_kwargs)
...
+
+A 4D mask must use one of two value conventions.
+
+| dtype | attend | mask out |
+|---|---|---|
+| boolean | `True` | `False` |
+| float | `0.0` | `-inf` |
```
Comment (vasqu) [inline docs/source/en/attention_interface.md]: Yesn't, this is only true for sdpa attention -> this one can handle bool and float masks; eager can handle float masks only. The rest needs to do their own thing

File: docs/source/en/attention_interface.md [hunk 3]
```diff
@@ -307,6 +307,57 @@ attention_mask = create_causal_mask(**mask_kwargs)
...
+
+A common mistake is to reuse the `1`/`0` convention of a 2D padding mask in a float 4D mask. Because the mask is added to the scores, `0.0` keeps a position and `1.0` only adds a small bias.
+
+The example below contrasts a wrong mask with a correct one. Both start from the same `1`/`0` causal pattern.
+
+```py
+import torch
+from transformers import AutoModelForCausalLM, AutoTokenizer
+
+model = AutoModelForCausalLM.from_pretrained("TinyLlama/TinyLlama-1.1B-Chat-v1.0")
```
Comment (vasqu) [inline docs/source/en/attention_interface.md]: Let's add the attn_implementation to highlight which attn backend it affects

**Review by vasqu — APPROVED** [review]
1 important thing to correct re when float/bool is ok but the rest is very neat, thanks!
